### PR TITLE
Give recovery behaviours a method to access to current local or global planner.

### PR DESCRIPTION
**Pull request by jasonimercer**
@skaynama  @ayrton04  @p6chen

### CODE REVIEW

**Comment by jasonimercer** [comment]
There is coupled with a branch of cpr_pathtracker with the same name.

File: move_base/include/move_base/move_base.h [hunk 1]
```diff
@@ -200,6 +200,18 @@ namespace move_base {
       boost::shared_ptr<nav_core::BaseGlobalPlanner> getGlobalPlannerPlugin(std::string plugin_name);
 
       /**
+       * @brief Used to retrieve the current global planner plugin.
+       * @return Pointer to the local planner plugin instance.
```
Comment (ayrton04) [inline move_base/include/move_base/move_base.h]: *global planner plugin instance

Comment (jasonimercer) [inline move_base/include/move_base/move_base.h]: too much copy and paste. Thanks for the spot.

**Comment by jasonimercer** [comment]
The reason I'm not just handing the planner plugins to the behaviours is that the planner can change. Providing a fetch function gets around this issue.

**Comment by ayrton04** [comment]
You answered the question before I could ask it. LGTM!

**Comment by jasonimercer** [comment]
Also using explicit setters rather than overloading the initialize with default empty fetch functions because the base class would have:

void initialize(foo bar, fetch1 f1 = empty, fetch2 f2 = empty) = 0;

and that secretly gets expanded to 

void initialize(foo bar, fetch1 f1, fetch2 f2) = 0;
void initialize(foo bar, fetch1 f1, empty) = 0;
void initialize(foo bar, empty, empty) = 0;

but the child classes don't implement all 3 required versions since they don't have default variables and so we are not allowed to instantiate them.

**Comment by skaynama** [comment]
Nice. LGTM.

**Comment by jasonimercer** [comment]
@p6chen , any objections to merging this PR?

**Comment by p6chen** [comment]
lgtm merge away

**Pull request merged**